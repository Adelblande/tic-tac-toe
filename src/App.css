#root {
  max-width: 1280px;
  margin: 0 auto;
  padding: 2rem;
  text-align: center;
}

div.board {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  gap: 1rem;
}

div.board span {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 100px;
  height: 100px;
  font-size: 2rem;
  background-color: #333;
  color: #fff;
  cursor: pointer;
}
