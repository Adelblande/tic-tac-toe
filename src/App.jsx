import "./App.css";
import { ResetButton, Timer, Board, Scores } from "./components/index.js";
import { useGame } from "./hooks/use-game";

function App() {
  const { game, message, player, scores, handleClick, resetGame, setPlayer } =
    useGame();

  return (
    <>
      <header className="header">
        <Scores scores={scores} />
        {!message && <Timer player={player} setPlayer={setPlayer} />}
      </header>

      <main className="main">
        {message && <span className="message">{message}</span>}
        <Board game={game} handleClick={handleClick} />
        <ResetButton onClick={resetGame} value="Reset Game" />
      </main>
    </>
  );
}

export default App;
