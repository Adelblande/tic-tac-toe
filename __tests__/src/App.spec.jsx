import { describe, it, expect } from "vitest";
import { render, waitFor } from "@testing-library/react";
import { screen } from "@testing-library/dom";
import userEvent from "@testing-library/user-event";
import App from "../../src/App";
import "@testing-library/jest-dom/vitest";

const user = userEvent.setup({ delay: null });

const executeXWinner = async () => {
  const cellX0 = screen.getAllByTestId("span-cell-0");
  await user.click(cellX0[0]);

  const cellO3 = screen.getAllByTestId("span-cell-3");
  await user.click(cellO3[0]);

  const cellX1 = screen.getAllByTestId("span-cell-1");
  await user.click(cellX1[0]);

  const cellO4 = screen.getAllByTestId("span-cell-4");
  await user.click(cellO4[0]);

  const cellX2 = screen.getAllByTestId("span-cell-2");
  await user.click(cellX2[0]);
};

const executeOWinner = async () => {
  const cellX2 = screen.getAllByTestId("span-cell-2");
  await user.click(cellX2[0]);

  const cellO1 = screen.getAllByTestId("span-cell-1");
  await user.click(cellO1[0]);

  const cellX3 = screen.getAllByTestId("span-cell-3");
  await user.click(cellX3[0]);

  const cellO4 = screen.getAllByTestId("span-cell-4");
  await user.click(cellO4[0]);

  const cellX5 = screen.getAllByTestId("span-cell-5");
  await user.click(cellX5[0]);

  const cellO7 = screen.getAllByTestId("span-cell-7");
  await user.click(cellO7[0]);
};

const executeCotOld = async () => {
  const cellX2 = screen.getAllByTestId("span-cell-2");
  await user.click(cellX2[0]);

  const cellO1 = screen.getAllByTestId("span-cell-1");
  await user.click(cellO1[0]);

  const cellX3 = screen.getAllByTestId("span-cell-3");
  await user.click(cellX3[0]);

  const cellO0 = screen.getAllByTestId("span-cell-0");
  await user.click(cellO0[0]);

  const cellX4 = screen.getAllByTestId("span-cell-4");
  await user.click(cellX4[0]);

  const cellO5 = screen.getAllByTestId("span-cell-5");
  await user.click(cellO5[0]);

  const cellX7 = screen.getAllByTestId("span-cell-7");
  await user.click(cellX7[0]);

  const cellO6 = screen.getAllByTestId("span-cell-6");
  await user.click(cellO6[0]);

  const cellX8 = screen.getAllByTestId("span-cell-8");
  await user.click(cellX8[0]);
};

describe("App", () => {
  it("should render screen of game", () => {
    const { container } = render(<App />);
    expect(container).toMatchSnapshot();
  });

  it("should return player x as winner", async () => {
    render(<App />);
    await executeXWinner();

    const message = screen.getByText("Player x won!");
    await waitFor(() => {
      expect(message).toBeInTheDocument();
    });
  });

  it("should return player o as winner", async () => {
    render(<App />);
    await executeOWinner();

    const message = screen.getByText("Player o won!");
    await waitFor(() => {
      expect(message).toBeInTheDocument();
    });
  });

  it("should remove timer of screen when has winner", async () => {
    render(<App />);
    await executeXWinner();

    const timer = screen.queryByTestId("timer");
    await waitFor(() => {
      expect(timer).not.toBeInTheDocument();
    });
  });

  it("should show scores acumulated", async () => {
    render(<App />);
    const reset = screen.getByRole("button", { name: "Reset Game" });
    await executeXWinner();
    await user.click(reset);
    await executeXWinner();
    await user.click(reset);
    await executeXWinner();

    await waitFor(() => {
      const scoreX = screen.getByText("Player X: 3");
      expect(scoreX).toBeInTheDocument();
    });
  });

  it("should show message Got old", async () => {
    render(<App />);
    await executeCotOld();

    const message = screen.getByText("Got old");
    await waitFor(() => {
      expect(message).toBeInTheDocument();
    });
  });

  it("should show a float menu to change colors of game", async () => {
    render(<App />);
    const button = screen.getByTestId("settings-button");
    await user.click(button);

    await waitFor(() => {
      const buttonsColors = screen.getAllByTestId(/settings-button-color-/);
      expect(buttonsColors.length).toEqual(3);
    });
  });
});
