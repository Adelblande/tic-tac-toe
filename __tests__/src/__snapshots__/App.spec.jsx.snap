// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App > should render screen of game 1`] = `
<div>
  <header
    class="_header_243a2a"
  >
    <div
      class="_scores_afa92b"
    >
      <h3>
        Scores
      </h3>
      <p>
        Player X: 
        0
      </p>
      <p>
        Player O: 
        0
      </p>
    </div>
    <div
      class="_timer_32e71e"
      data-testid="timer"
    >
      <h3>
        Player 
        X
        : 
        5
      </h3>
    </div>
  </header>
  <main
    class="_main_6f6c26"
    style="background-color: rgb(214, 193, 239);"
  >
    <div
      class="_board_f7a227"
    >
      <button
        data-testid="span-cell-0"
        id="cell-0"
        type="button"
      />
      <button
        data-testid="span-cell-1"
        id="cell-1"
        type="button"
      />
      <button
        data-testid="span-cell-2"
        id="cell-2"
        type="button"
      />
      <button
        data-testid="span-cell-3"
        id="cell-3"
        type="button"
      />
      <button
        data-testid="span-cell-4"
        id="cell-4"
        type="button"
      />
      <button
        data-testid="span-cell-5"
        id="cell-5"
        type="button"
      />
      <button
        data-testid="span-cell-6"
        id="cell-6"
        type="button"
      />
      <button
        data-testid="span-cell-7"
        id="cell-7"
        type="button"
      />
      <button
        data-testid="span-cell-8"
        id="cell-8"
        type="button"
      />
    </div>
    <button
      class="_resetButton_baa250"
    >
      Reset Game
    </button>
    <div
      class="_set_ab705f"
    >
      <button
        class="_mainMenu_ab705f"
        data-testid="settings-button"
      >
        <img
          alt="Menu settings"
          height="30"
          src="/tic-tac-toe/settings.png"
          width="30"
        />
      </button>
      <ul>
        <li>
          <button
            data-testid="settings-button-color-0"
          >
            <span
              style="color: rgb(127, 0, 164);"
            >
              x
            </span>
            <span
              style="color: rgb(208, 8, 241);"
            >
              o
            </span>
            <span
              style="color: rgb(214, 193, 239);"
            >
              #
            </span>
          </button>
        </li>
        <li>
          <button
            data-testid="settings-button-color-1"
          >
            <span
              style="color: rgb(0, 130, 9);"
            >
              x
            </span>
            <span
              style="color: rgb(80, 204, 86);"
            >
              o
            </span>
            <span
              style="color: rgb(40, 167, 48);"
            >
              #
            </span>
          </button>
        </li>
        <li>
          <button
            data-testid="settings-button-color-2"
          >
            <span
              style="color: rgb(139, 0, 4);"
            >
              x
            </span>
            <span
              style="color: rgb(225, 25, 64);"
            >
              o
            </span>
            <span
              style="color: rgb(185, 10, 30);"
            >
              #
            </span>
          </button>
        </li>
      </ul>
    </div>
  </main>
</div>
`;
